hudi-flink/src/main/java/org/apache/hudi/sink/partitioner/BucketAssignFunction.java

```diff
@@ -90,7 +92,7 @@
    *   <li>If it does not, use the {@link BucketAssigner} to generate a new bucket ID</li>
    * </ul>
    */
-  private MapState<HoodieKey, HoodieRecordLocation> indexState;
+  private MapState<String, GlobalHoodieRecordLocation> indexState;
```

Comment:
If we need to support a global index, we need to use a single source across parallelisms as the index store right? How does this distributed state ensure the consistency of the result? Let's say for the case that we change the parallelism and restart the job. The keys and parallelism hash mapping were reorganized.

Reply:
No, the parallelism does not affect the semantics. We do deduplication based on record keys and the global hoodie record location can recognize the record location among partitions.

Reply:
Let me give an example. When a record changes the partition, we will create a delete record to the old location and update the index with the new location, but this index state was globally distributed and only one parallelism will be affected. If we change the parallelism and restart the job, this record may be in different parallelism with a wrong location in the index state.

Reply:
Also, the global index is very expensive for a large table. Can we make the index type configurable?

Reply:
> Let me give an example. When a record changes the partition, we will create a delete record to the old location and update the index with the new location, but this index state was globally distributed and only one parallelism will be affected. If we change the parallelism and restart the job, this record may be in different parallelism with a wrong location in the index state.

No worries, the flink state backed would keep the semantics for us.

Reply:
> Also, the global index is very expensive for a large table. Can we make the index type configurable?

Fine with that.

Reply:
> > Let me give an example. When a record changes the partition, we will create a delete record to the old location and update the index with the new location, but this index state was globally distributed and only one parallelism will be affected. If we change the parallelism and restart the job, this record may be in different parallelism with a wrong location in the index state.
> 
> No worries, the flink state backed would keep the semantics for us.

That's interesting! I just started to learn Flink and found that the keyedState was only responsible for local change. If we change the parallelism, the data layout should be completely changed since the keyBy operation uses a different parallelism number. Does Flink automatically sync the distributed state for this case? Any doc for this feature?